Firestore/core/src/firebase/firestore/immutable/llrb_node.h

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_LLRB_NODE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_LLRB_NODE_H_
+
+#include <memory>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/immutable/sorted_map_base.h"
+
+namespace firebase {
+namespace firestore {
+namespace immutable {
+namespace impl {
+
+/**
+ * A Color of a tree node in a red-black tree.
+ */
+enum Color {
+  Black = 0u,
+  Red = 1u,
+
+  Default = static_cast<unsigned int>(Red),
+};
+
+/**
+ * LlrbNode is a node in a TreeSortedMap.
+ */
+template <typename K, typename V>
+class LlrbNode : public SortedMapBase {
+ public:
+  using first_type = K;
+  using second_type = V;
+
+  /**
+   * The type of the entries stored in the map.
+   */
+  using value_type = std::pair<K, V>;
+
+  /**
+   * Constructs an empty node.
+   */
+  LlrbNode()
+      : LlrbNode(NodeData{std::pair<K, V>{}, Color::Black, /*size=*/0u,
```

Comment:
* `std::pair<K, V>{}` can be just `{}` (optional)
* perhaps add a constructor to `NodeData` instead? Note that `NodeData` is, AFAICS, neither trivial (because it stores `shared_ptr` nor default layout (because of the bit field), so I don't think there's a restriction on that.